### DISCUSSION TITLE
Stratified Flows module

### DISCUSSION THREAD

**Discussion by chowland** [FourierFlows/GeophysicalFlows.jl#331] [Ideas]
We've discussed this before, but now I've submitted my thesis (yay!), I actually have time to start working on a module for a vertical plane with an imposed mean stratification. There are a couple of options for the implementation of this:

- Vorticity-streamfunction formulation. Should be relatively simple to implement given its similarity to the existing modules.
- Primitive variable formulation for velocity (_u_, _w_) and buoyancy perturbation _θ_ with a pressure solve through a projection method. I saw you have a 3D box as a potential future development on FourierFlows, so implementing a pressure solver may be useful to lead in to that.

Any thoughts?

Once this is working it should be relatively simple to include multiple scalars and make some cool examples of salt fingering and diffusive convection.

I'll be using the existing modules as a starting point and will let you know how I get on. Any tips for developing these modules are much appreciated!

**Comment by navidcy** [comment]
hey @chowland congrats on submitting thesis!
A stratified flows module would be great.
FourierFlows.jl already includes  a `ThreeDGrid` but we don't have any module in GeophysicalFlows.jl using it. 
Can you point to a pdf/notes with the equations written down so I think about it a bit?

**Comment by chowland** [comment]
Thanks @navidcy !
Here's a document going over the two options in a bit more detail:
[stratified_periodic_equations.pdf](https://github.com/FourierFlows/GeophysicalFlows.jl/files/4824380/stratified_periodic_equations.pdf)

The method I suggested for the pressure is what Diablo uses, and may come in useful if you want to make a 3DNavierStokes module.
I've also outlined a couple of extensions that it would be good to have in mind while developing this.

**Comment by glwagner** [comment]
@chowland nice! Congrats!

I think a 2D vorticity-streamfunction formulation might be a good place to start because of its simplicity.

However, one consideration is that a 3D implementation could be easily reduced to 2D by setting, for example, `ny=1`. The downside is the added complexity. The upside is that we immediately obtain efficient 3D and 2D solvers.

As for the pressure solve, we do not need to use pressure projection or fractional time stepping methods with a 3D spectral method. Instead, we can obtain the pressure directly from the transformed equations in Fourier space, and substitute the result into the RHS. This approach is described in [Winters et al 2004](https://watermark.silverchair.com/1520-0426(2004)021_0069_asmfps_2_0_co_2.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAAtwwggLYBgkqhkiG9w0BBwagggLJMIICxQIBADCCAr4GCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMo_fHx3IJxUHjQPqmAgEQgIICj8y8TDb88OaeR_YWghtHHN26zYA6RAvAaESwoHrmkPetqynhLQRE6FGkuAEW-0r6it4nl8iNBjx4i90tkBB-7W_ZWO98p2vvY3uUmPG0I5muDJ1jprogIngtdx8dAQepKwCLcdU_fSHIjiCjfS6hLfe6mPmYk92wqCoReMbohTbr3zjMTlOcPAyDZ2WMS9YbkR1_uCG5wp-j8-9FNkjg4jy_KCTJvp4DlrogTOlmYRLr6trGD2YiAEWDM5bid7WrMnx0EmbyGCpb784AWvQS6Thsa2v9DJLC5D17JgsofB45ckrYl83ITYJWZMs6es9X4-0NF3P_utz3F8y_ah5eLvgm6QjGTQg7JNl9sXxIoT5ldNkPkEKOIZB_tmFUp7WXgs9AgbnBfbCDrFrMTXDOz4bUPhxcneNfcS02MawL7J5lacxGMhjCYzN0UhEPlLEwrEU58RYzT_D0N-nzQXveLTY4iMmXBBO8MdqoEk0ovTVxMtAcHFACYZcrxqrA0-8SPavgQic-1qbclJFV-5aodGuGS6iExODV-Sh6m3X_VsSRuhaWIxD0oESqnpcEWPDaEeLtYA0B6AL_BqQ-LCbBgvw8PFkoVeoWRhFUT6ef5lBMwWMVB9Ykn0AqnGEnTKD9XihRcC0V1-Ay4n1iYW3Y5NOajqaChQUyGpE0xhdKL5xJjNSD3ruPv2_avtwovcziNxFzXlaVGiFd9k029xAQC4KmaYdDXsUFxgpLl0DyLjoq3tlQGL37ty8o5EGXfWBH5qrqMYM5d235VniW_fxkV6casyi0BPIXTUPAPPCkbloA9Td1QZcV_awgPssSLRl82GU5h8UyxC2-hSOwUfji6LoEkHHV-FHM9Rmt1SU_Ao0) (see equations 33-35).

Looking at those equations, I do think it probably makes sense to start with the vorticity-streamfunction formulation. It's much simpler.

I'd suggest using `b` rather than `θ` for buoyancy. `θ` is often used for "temperature".

Finally, we should probably also have rotation! This requires an additional equation for `v` in the vorticity-streamfunction formulation. The equations are given in [Xie and Buhler (2019)](https://www.researchgate.net/profile/Jin_Han_Xie/publication/333795799_Two-dimensional_isotropic_inertia-gravity_wave_turbulence/links/5d75b9ef4585151ee4a8bf02/Two-dimensional-isotropic-inertia-gravity-wave-turbulence.pdf). We can probably implement a rotating version as an "option" in the new module (we don't need to put rotation in at the outset).
